### PR TITLE
(stm32) RIF configuration fixes and device tree update for STM32MP257F-EV1 board

### PR DESCRIPTION
**Pull request by GseoC**
This P-R updates and fixes the RIF configuration application in some RIF-aware peripherals and fixes the generic parsing.
Those RIF-aware peripherals were not enabled for now because we lacked clock and reset support so enable most of them now.

### CODE REVIEW

File: core/drivers/stm32_ipcc.c [hunk 1]
```diff
@@ -217,7 +220,6 @@ stm32_ipcc_pm(enum pm_op op, unsigned int pm_hint,
 		panic();
 
 	if (!PM_HINT_IS_STATE(pm_hint, CONTEXT) || !is_tdcid)
-		return TEE_SUCCESS;
```
Comment (etienne-lms) [inline core/drivers/stm32_ipcc.c:238]: Restore the return instruction (see commit "drivers: stm32_ipcc: do not make RIF configuration mandatory").

**Review by etienne-lms — COMMENTED** [review]
`Reviewed-by: Etienne Carriere <etienne.carriere@foss.st.com>` for commits
"drivers: stm32_hpdma: do not make RIF configuration mandatory" (with comment addressed),
"drivers: stm32_fmc: use PM_HINT_IS_STATE macro in pm callbacks",
"drivers: stm32_hsem: fix unfreed memory" and 
"dts: stm32: update some RIF-aware peripherals nodes in stm32mp251".

File: core/drivers/stm32_hpdma.c [hunk 2]
```diff
@@ -179,18 +184,26 @@ static TEE_Result apply_rif_config(struct hpdma_pdata *hpdma_d, bool is_tdcid)
...
 end:
 	if (IS_ENABLED(CFG_TEE_CORE_DEBUG)) {
 		/* Check that RIF config are applied, panic otherwise */
 		if ((io_read32(hpdma_d->base + _HPDMA_PRIVCFGR) &
 		     hpdma_d->conf_data->access_mask[0]) !=
-		    hpdma_d->conf_data->priv_conf[0])
-			panic("HPDMA channel priv conf is incorrect");
+		    hpdma_d->conf_data->priv_conf[0]) {
+			EMSG("HPDMA channel priv conf is incorrect");
+			panic();
```
[the remark this answers is not in the document]
Comment (GseoC) [reply core/drivers/stm32_hpdma.c:198]: no :)

File: core/drivers/stm32_fmc.c [hunk 3]
```diff
@@ -244,28 +246,33 @@ static TEE_Result parse_dt(const void *fdt, int node)
...
+	fmc_d->conf_data->priv_conf = calloc(1, sizeof(uint32_t));
+	fmc_d->conf_data->lock_conf = calloc(1, sizeof(uint32_t));
+	fmc_d->conf_data->access_mask = calloc(1, sizeof(uint32_t));
+	if (!fmc_d->conf_data->cid_confs || !fmc_d->conf_data->sec_conf ||
+	    !fmc_d->conf_data->priv_conf || !fmc_d->conf_data->access_mask)
+		panic("Missing memory capacity for FMC RIF configuration");
 
-	for (i = 0; i < fmc_d->nb_controller; i++) {
-		rif_conf = fdt32_to_cpu(cuint[i]);
+	stm32_rif_parse_cfg(fdt32_to_cpu(cuint[i]), fmc_d->conf_data,
```
[the remark this answers is not in the document]
Comment (GseoC) [reply core/drivers/stm32_fmc.c:275]: I do, thanks

**Comment by GseoC** [comment]
Comments addressed. Added a fix in the stm32_fmc driver as well, commit: `
drivers: stm32_fmc: fix check_fmc_rif_conf() and its usage`

File: core/include/drivers/stm32_rif.h [hunk 4]
```diff
@@ -135,7 +129,6 @@ TEE_Result stm32_rif_check_access(uint32_t cidcfgr,
  */
 void stm32_rif_parse_cfg(uint32_t rif_conf,
 			 struct rif_conf_data *conf_data,
-			 unsigned int nb_cid_supp,
 			 unsigned int nb_channel);
```
Comment (etienne-lms) [inline core/include/drivers/stm32_rif.h:132]: Remove inline comment at line 127.

**Review by etienne-lms — COMMENTED** [review]
Most of the commits are fixes hence deserve a `Fixes:` tag in their commit message:
* commit "drivers: stm32_rif: update and fix RIF configuration parsing":
`Fixes: 1506f47af917 ("drivers: firewall: add stm32_rif driver for common RIF features")`
* commit "drivers: stm32_hpdma: update and fix CID filtering configuration":
`Fixes: 0cf1cd13ceb6 ("drivers: add HPDMA driver with RIF support")`
* commit "drivers: stm32_fmc: update and fix CID filtering configuration":
`Fixes: 778a36bf91d2 ("drivers: add FMC driver with RIF support")`
* commit "drivers: stm32_ipcc: update and fix RIF configuration":
`Fixes: e26b8e0f3e4c ("drivers: add IPCC driver and its RIF support")`
* commit "drivers: stm32_hsem: fix unfreed memory":
`Fixes: ec9aa1a4114f ("drivers: add RIF support driver for HSEM")`
* commit "drivers: stm32_hsem: fix HSEM group CID filtering configuration":
`Fixes: ec9aa1a4114f ("drivers: add RIF support driver for HSEM")`
* commit "drivers: stm32_gpio: update and fix RIF configuration handling":
`Fixes: bd03c8c3d70f  ("drivers: stm32_gpio: add stm32mp25x support")`
* commit "dts: stm32: fix stm32mp257f-ev1 clock tree configuration":
`Fixes: 9223d8a0fc1d ("dts: st: add RCC support on stm32mp257f-ev1")`
* commit "drivers: stm32_fmc: fix check_fmc_rif_conf() and its usage":
`Fixes: 778a36bf91d2 ("drivers: add FMC driver with RIF support")`

Can you fix and squash all fixup commits. I'll post some review tags afterward. Most changes LGTM,  I need to spend at bit more time on the hpdma/fmc/ipcc/hsem "update and fix" commits.

File: core/drivers/stm32_fmc.c [hunk 5]
```diff
@@ -316,8 +316,10 @@ static TEE_Result __maybe_unused check_fmc_rif_conf(void)
 	if (res)
 		panic("Cannot access FMC clock");
 
-	if (fmc_controller_is_secure(0))
+	if (fmc_controller_is_secure(0)) {
+		res = TEE_SUCCESS;
```
[the remark this answers is not in the document]
Comment (GseoC) [reply core/drivers/stm32_fmc.c:320]: Keep as is then, I found it better to explicit :)

File: core/drivers/stm32_gpio.c [hunk 6]
```diff
@@ -785,42 +817,21 @@ static TEE_Result apply_rif_config(struct stm32_gpio_bank *bank)
...
 
 	/*
 	 * Lock RIF configuration if configured. This cannot be undone until
 	 * next reset.
 	 */
-	io_setbits32(bank->base + GPIO_RCFGLOCKR_OFFSET,
-		     bank->rif_cfg->lock_conf[0]);
+	io_clrsetbits32(bank->base + GPIO_RCFGLOCKR_OFFSET,
+			bank->rif_cfg->access_mask[0],
+			bank->rif_cfg->lock_conf[0]);
```
[the remark this answers is not in the document]
Comment (GseoC) [reply core/drivers/stm32_gpio.c:828]: Propagated to FMC and HPDMA drivers

**Comment by GseoC** [comment]
Comments addressed

**Comment by GseoC** [comment]
Updated to address code style issue

File: core/drivers/stm32_hpdma.c [hunk 7]
```diff
@@ -143,41 +174,18 @@ static TEE_Result apply_rif_config(struct hpdma_pdata *hpdma_d, bool is_tdcid)
...
 	}
 
 	/*
 	 * Lock RIF configuration if configured. This cannot be undone until
 	 * next reset.
 	 */
-	io_clrsetbits32(hpdma_d->base + _HPDMA_RCFGLOCKR, _HPDMA_RCFGLOCKR_MASK,
-			hpdma_d->conf_data->lock_conf[0]);
+	io_setbits32(hpdma_d->base + _HPDMA_RCFGLOCKR,
+		     hpdma_d->conf_data->lock_conf[0] & _HPDMA_RCFGLOCKR_MASK);
```
Comment (etienne-lms) [inline core/drivers/stm32_hpdma.c:184]: `& _HPDMA_RCFGLOCKR_MASK` not needed.
Ditto for the stm32_fmc and stm32_gpio driver patches.

**Review by etienne-lms — COMMENTED** [review]
`Reviewed-by: Etienne Carriere <etienne.carriere@foss.st.com>` for commits
"drivers: stm32_rif: update and fix RIF configuration parsing",
"drivers: stm32_hpdma: do not make RIF configuration mandatory",
"drivers: stm32_fmc: do not make RIF configuration mandatory",
"drivers: stm32_fmc: fix check_fmc_rif_conf() and its usage" (by the way, I think you should use "use" instead of "usage" in the commit message header line),
"drivers: stm32_ipcc: do not make RIF configuration mandatory" and
"dts: stm32: fix stm32mp257f-ev1 clock tree configuration".

`Acked-by: Etienne Carriere <etienne.carriere@foss.st.com>` for commit
"dts: stm32: support some RIF-aware peripherals".

Remains 5 commits I need to go through again.

File: core/drivers/stm32_fmc.c (hunk 3 above)
[the remark this answers is not in the document]
Comment (etienne-lms) [reply core/drivers/stm32_fmc.c]: You forgot to restore the `for()` loop in commit "drivers: stm32_fmc: do not make RIF configuration mandatory".

File: core/drivers/stm32_gpio.c [hunk 8]
```diff
@@ -1312,6 +1326,7 @@ static TEE_Result apply_sec_cfg(void)
 				     'A' + bank->bank_id);
 				STAILQ_REMOVE(&bank_list, bank, stm32_gpio_bank,
 					      link);
+				free(bank);
```
Comment (etienne-lms) [inline core/drivers/stm32_gpio.c:1329]: Could you also mention this fix in the commit message body?

File: core/drivers/stm32_fmc.c (hunk 3 above)
Comment (etienne-lms) [inline core/drivers/stm32_fmc.c]: You forgot to restore the `for()` loop in commit "drivers: stm32_fmc: do not make RIF configuration mandatory".

**Review by etienne-lms — COMMENTED** [review]
`Acked-by: Etienne Carriere <etienne.carriere@foss.st.com>` for commits
"drivers: stm32_hpdma: update and fix CID filtering configuration",
"drivers: stm32_ipcc: update and fix RIF configuration",
"drivers: stm32_hsem: fix HSEM group CID filtering configuration" and 
"drivers: stm32_gpio: update and fix RIF configuration handling" (with commit message updated, see my review comment).

Commit "drivers: stm32_fmc: update and fix CID filtering configuration" LGTM but depends on the its preceding commit that requires an update (see comment on missing `for()` loop).

**Comment by GseoC** [comment]
Tags applied and comments addressed, 1 commit remaining

File: core/drivers/stm32_gpio.c (hunk 8 above)
Comment (etienne-lms) [reply core/drivers/stm32_gpio.c:1329]: Don't forget to update the commit message.

**Review by etienne-lms — COMMENTED** [review]
`Acked-by: Etienne Carriere <etienne.carriere@foss.st.com>` for commit
"drivers: stm32_fmc: update and fix CID filtering configuration".

File: core/drivers/stm32_gpio.c (hunk 8 above)
Comment (GseoC) [reply core/drivers/stm32_gpio.c:1329]: Yes, done, thanks for the remidner

**Comment by GseoC** [comment]
Last tag applied and comment added in `drivers: stm32_gpio: update and fix RIF configuration handling` commit message, thanks

**Pull request merged**